### DISCUSSION TITLE
How do I convert an existing repository to git lfs

### DISCUSSION THREAD

**Discussion by TG-009** [git-lfs/git-lfs#5235] [General]
I am sure this has been answered however I could not find any discussion about this.
I have an existing repository with many branches etc. that we have been using for several years.
We have now discovered that several large files are not being managed by git (i.e. they are not in the repository)
So we want to use git-lfs
I have read the various documentation on setting up git-lfs and I believe I have created a brand new blank repository that is set up for git-lfs.  In essence it has a .gitattribute and a .gitignore file that is all.
So I asked a team mate to move our project to this new repository.
We now have a new branch on this repository with our project files on it.  I am using Git Desktop.  When I try to change to this branch I am getting errors  - see below
"Updating files: 100% (6816/6816), done.
Downloading Assets/Editor Default Resources/FMOD/AddIcon.png (380 B)
Error downloading object: Assets/Editor Default Resources/FMOD/AddIcon.png (860e3a6): Smudge error: Error downloading Assets/Editor Default Resources/FMOD/AddIcon.png (860e3a60f97b4f9961eda80617e922ceffeb67aa145077ba0599863ca306a467): [860e3a60f97b4f9961eda80617e922ceffeb67aa145077ba0599863ca306a467] Object does not exist on the server: [404] Object does not exist on the server.
Errors logged to 'E:\T Documents\Unity Projects\_ProjectsFolder\MyGitHubRepos\VL_LFS\.git\lfs\logs\20221220T175601.447822.log'.
Use `git lfs logs last` to view the log.
error: external filter 'git-lfs filter-process' failed
fatal: Assets/Editor Default Resources/FMOD/AddIcon.png: smudge filter lfs failed"

Now, how are we supposed to take my existing project and transfer it to this new repository??

**Comment by chrisd8088** [comment]
The error message `Object does not exist on the server: [404] Object does not exist on the server` indicates that the Git LFS object was not pushed to the server, and is therefore not available to be fetched.

I would suggest that you start by working with the copy of your new repository which has all the large files added to it locally, so perhaps your colleague's copy.  In other words, start with the copy into which the files were moved and committed.  They should appear when you run `git lfs ls-files`.  Then push them with `git push` or `git lfs push` to make sure the Git LFS objects are transferred to the server.  After that, other people should be able to fetch them.

Atlassian has a good introductory [tutorial](https://www.atlassian.com/git/tutorials/git-lfs) into using Git LFS.

If you are adding Git LFS to an existing repository, you may want to use `git lfs migrate import` and review its [manual page](https://github.com/git-lfs/git-lfs/blob/4140f638e9e305605f2e8cd941355340bb71fbf0/docs/man/git-lfs-migrate.adoc), but I would get familiar with basic Git LFS concepts first.

**Reply by TG-009** [reply]
Thank you for the answer.
I understand the concept of git lfs (large files stored elsewhere with pointers to them in our git project).  Other than that what more do I need to know??!!?
I do not want to change my existing git repository to git lfs for fear that I screw something up and destroy my project (4 years in the making).  My preference is to create a new git lfs repository and transfer my project to it as I feel that is safer  (That way if I screw something up I can go back to my old repository - call me crazy).
I do not use git terminal, or command line nor know what that is.  I use git desktop to make branches, pull and push that's it.  I really need some very clear non git technical instructions on 1.  creating a new git lfs remote repository on GitHub 2. How do I make the first push of my project to that remote repository  3. How do others on the team pull from that new remote repository.

I have read a lot of the git-lfs documentation but it quickly gets to be git technical beyond my capacity to understand.
Any direction you can give me is appreciated.

I read the Atlassian beginner tutorial you pointed me to, but it is all Atlassian & command line and I don't know how to do that.

I have installed git lfs on my computer and run install and it says it has done that correctly.  Not sure what to do next to solve this.

**Comment by chrisd8088** [comment]
Creating a new repository, if you're able to do that, is certainly a good idea and will indeed be a safe way to proceed.

Since this project is just for the open-source command-line Git LFS client, I'm not sure anyone will be able to provide good advice regarding the GitHub Desktop utility.  That uses Git LFS internally, but adds its own user interface on top.  I would suggest getting in touch with GitHub Support and ask them for help.

From what I found in their [documentation](https://docs.github.com/en/desktop/installing-and-configuring-github-desktop/configuring-and-customizing-github-desktop/about-git-large-file-storage-and-github-desktop), though, you may need to work with the command line to proceed:

> To use Git LFS with GitHub Desktop, you must configure Git LFS using the command line. For more information, see "[Configuring Git Large File Storage](https://docs.github.com/en/github/managing-large-files/configuring-git-large-file-storage)."

A few other points: Git LFS is generic and not specific to GitHub, Atlassian, or any other system.  That's why I suggested reading the tutorial, because it's generic enough to cover any Git hosting service.

Since you may have to work with the command line, I would suggest that the description in GitHub's [documentation](https://docs.github.com/en/repositories/working-with-files/managing-large-files/configuring-git-large-file-storage) is quite good.

In essence, if you have created an empty new repository, you would do something like the following:
- `cd` into your  newrepository's directory, which I'll assume you've created with GitHub Desktop
- run `git lfs track` for each file type (i.e., file extension) you want to track with Git LFS, e.g., `git lfs track "*.jpg"` or whatever you need
- that creates a `.gitattributes` file (note the plural); I believe you said you've already done this
- run `git add .gitattributes` to make sure it will be included in your repository's contents
- copy in all the files you want in your repository from your old repository
- run `git add *` so they will all be included in the repository's contents
- run `git commit`

Now you should have all your files in a first commit, with any that match the rules you created with `git lfs track` converted into Git LFS objects.

At this point, you should be able to use GitHub Desktop to push the repository, but on the command line it would not be too difficult either, assuming you have set up the Git remote you want.  I would use `git remote -v` to see which remote servers you have configured (likely `origin`), and then `git push <remote> <branch>`, e.g., `git push origin main`.

**Reply by TG-009** [reply]
Thank you for the clarification and clear instructions.
Here is what I plan to do and if you would be so kind as to let me know if I have this correct.

I will not create a new lfs remote repository on my Github account ahead of time as I believe step 6 below will do that for me.
1.  On my local machine I have created a new folder and copied my project into that folder with all of my large files
2.   I have gone to that new folder and "right click", have run "Git Bash here"
3. Using git bash entered "git lfs install", with the result being "Updated Git hooks." and "Git LFS initialized."
4. I have done "git lfs track"  multiple times to create the .gitattributes file that I want
5. I could follow your instructions for "git add .gitattributes" thru to "git commit" however my thought was to use GitHub desktop (as it is already configured to my remote Github account) and do "File / Add local Repository" with the file location being my new folder and then do a commit.
6. Then a "Publish repository"

If I understand git properly this will create a remote repository on Github in the lfs style and my project will be in that root location.

Other team members can then just pull, push, branch normally and LFS will be working in the background.  By LFS working in the background I mean that if any new large files are needed on any team members local machine they are copied to that local machine and if any large files are added by team members and they push, then those large files appear on the remote github (as pointers to some other git large file storage location).

How am I doing, do I understand things correctly??

Thank you for your time on this.

**Comment by chrisd8088** [comment]
That sounds like it might work, but I confess that I have no experience with GitHub Desktop, so you'll have to just try it and ask the GitHub Support people for help if it doesn't.

And I'm sorry but I don't know if GitHub Desktop supports Git LFS transparently the way you describe.  I would guess so, based on their documentation, but I really don't know.  Please report back what your experience is, and what the GitHub Support people tell you.  Good luck!

**Reply by TG-009** [reply]
chrisd8088
I followed the instructions as I laid them out above and when I added the new local repository, GitHub Desktop popped up this message "Initialize Git LFS, This repository uses Git LFS.  To contribute to it, Git LFS must first be initialized.  would you like to do  so now?"  So I selected to "Initialize Git LFS"  (even though I had already done that).
I then committed to main which took several minutes and then published.
It all worked well.
![Git-LFS message](https://user-images.githubusercontent.com/51301955/210103703-5006907a-ad96-49c6-9da7-4de351c9b346.png)

Thank you for help with this.

**Comment by TG-009** [comment]
Well, just to keep this post complete.
As mentioned above I pushed my project to the remote github and even though it appeared that it worked. It may not have.
Another person on the team tried to pull from the LFS repository and it did not work.  The reason we think it did not work is because he did not receive the large files.  The file names do appear on his local machine, but they are binary files not the actual large file (a music wav file) !!!
So we are back to trying to figure out what went wrong.
Any thoughts by anyone would be appreciated.
